Embedded struct fields are not supported by compiler

## Is your feature request related to a problem? Please describe.
I'm always frustrated when I look at the code generated for embedded fields. So I try testing it and for
```
--- a/pkg/compiler/struct_test.go
+++ b/pkg/compiler/struct_test.go
@@ -397,6 +397,18 @@ var structTestCases = []testCase{
                `,
                big.NewInt(42),
        },
+       {
+               "embedded struct",
+               `type S11 struct { S10; b int }
+               func F%d() int {
+                       var s S11
+                       s.a = 31
+                       s.b = 11
+                       return s.a + s.b
+               }
+               `,
+               big.NewInt(42),
+       },
        {
                "omit field names",
                `type pair struct { a, b int }
```
I get
```
--- FAIL: TestStructs (0.02s)
    --- FAIL: TestStructs/embedded_struct (0.00s)
        vm_test.go:69: 
                Error Trace:    /home/rik/dev/neo-go/pkg/compiler/vm_test.go:69
                                                        /home/rik/dev/neo-go/pkg/compiler/struct_test.go:461
                Error:          Received unexpected error:
                                at instruction 499 (SETITEM): unhandled exception: "The value -1 is out of range."
                Test:           TestStructs/embedded_struct
FAIL
FAIL    github.com/nspcc-dev/neo-go/pkg/compiler        9.481s
FAIL
```

## Describe the solution you'd like
Add proper support for it.

--- a/pkg/compiler/analysis.go
+++ b/pkg/compiler/analysis.go
@@ -226,15 +226,26 @@ func isExprNil(e ast.Expr) bool {
 	return ok && v.Name == "nil"
 }
 
-// indexOfStruct returns the index of the given field inside that struct.
-// If the struct does not contain that field, it will return -1.
-func indexOfStruct(strct *types.Struct, fldName string) int {
+// pathToField returns a slice of indices (in reverse order: innermost to outermost)
+// locating fldName inside strct, or nil if not found.
+func pathToField(strct *types.Struct, fldName string) []int {
 	for i := range strct.NumFields() {
 		if strct.Field(i).Name() == fldName {
-			return i
+			return []int{i}
 		}
 	}
-	return -1
+	for i := range strct.NumFields() {
+		fld := strct.Field(i)
+		if !fld.Embedded() {
+			continue
+		}
+		if inner, ok := getStruct(fld.Type()); ok {
+			if p := pathToField(inner, fldName); p != nil {
+				return append(p, i)
+			}
+		}
+	}
+	return nil
 }
 
 type funcUsage map[string]bool

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -228,9 +228,12 @@ func (c *codegen) emitLoadConst(t types.TypeAndValue) {
 	}
 }
 
-func (c *codegen) emitLoadField(i int) {
-	emit.Int(c.prog.BinWriter, int64(i))
-	emit.Opcodes(c.prog.BinWriter, opcode.PICKITEM)
+// emitLoadField loads a field using the given path in reverse order from innermost to outermost.
+func (c *codegen) emitLoadField(path []int) {
+	for i := len(path) - 1; i >= 0; i-- {
+		emit.Int(c.prog.BinWriter, int64(path[i]))
+		emit.Opcodes(c.prog.BinWriter, opcode.PICKITEM)
+	}
 }
 
 func (c *codegen) emitStoreStructField(i int) {
@@ -252,9 +255,14 @@ func (c *codegen) emitStoreSelectorExpr(n *ast.SelectorExpr) {
 		c.prog.Err = fmt.Errorf("nested selector assigns not supported yet")
 		return
 	}
-	ast.Walk(c, n.X)                      // load the struct
-	i := indexOfStruct(strct, n.Sel.Name) // get the index of the field
-	c.emitStoreStructField(i)             // store the field
+	ast.Walk(c, n.X)                       // load the struct
+	path := pathToField(strct, n.Sel.Name) // get path to field
+	if path == nil {
+		c.prog.Err = fmt.Errorf("field %q not found in type %s", n.Sel.Name, typ)
+		return
+	}
+	c.emitLoadField(path[1:])       // load the field
+	c.emitStoreStructField(path[0]) // store the field
 }
 
 // emitStoreIndexExpr emits code to store into an index expression (container[index]).
@@ -1199,8 +1207,12 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 			return nil
 		}
 		ast.Walk(c, n.X) // load the struct
-		i := indexOfStruct(strct, n.Sel.Name)
-		c.emitLoadField(i) // load the field
+		path := pathToField(strct, n.Sel.Name)
+		if path == nil {
+			c.prog.Err = fmt.Errorf("field %q not found in type %s", n.Sel.Name, typ)
+			return nil
+		}
+		c.emitLoadField(path) // load the field
 		return nil
 
 	case *ast.UnaryExpr:
@@ -2011,11 +2023,11 @@ func (c *codegen) convertBuiltin(expr *ast.CallExpr) {
 				emit.Opcodes(c.prog.BinWriter, opcode.DUP, opcode.SIZE)   // x y cnt x y len(y)
 				emit.Opcodes(c.prog.BinWriter, opcode.PUSH3, opcode.PICK) // x y cnt x y len(y) cnt
 				after := c.newLabel()
-				emit.Jmp(c.prog.BinWriter, opcode.JMPEQL, after) // x y cnt x y
+				emit.Jmp(c.prog.BinWriter, opcode.JMPEQL, after)          // x y cnt x y
 				emit.Opcodes(c.prog.BinWriter, opcode.PUSH2, opcode.PICK, // x y cnt x y cnt
 					opcode.PICKITEM, // x y cnt x y[cnt]
 					opcode.APPEND,   // x=append(x, y[cnt]) y cnt
-					opcode.INC) // x y cnt+1
+					opcode.INC)      // x y cnt+1
 				emit.Jmp(c.prog.BinWriter, opcode.JMPL, start)
 				c.setLabel(after)
 				for range 4 { // leave x on stack

--- a/pkg/compiler/struct_test.go
+++ b/pkg/compiler/struct_test.go
@@ -494,6 +494,69 @@ var structTestCases = []testCase{
 		`,
 		big.NewInt(1),
 	},
+	{
+		"embedded struct",
+		`type S12 struct { b1 int; S10; b2 int; }
+		func F%d() int {
+			var s S12
+			s.a = 31
+			s.b1 = 11
+			return s.b1 + s.a
+		}
+		`,
+		big.NewInt(42),
+	},
+	{
+		"nested embedded",
+		`type S13 struct { c1 int; c2 int; S12 }
+		func F%d() int {
+			var s S13
+			s.a = 31
+			s.b1 = 11
+			return s.a + s.b1
+		}
+		`,
+		big.NewInt(42),
+	},
+	{
+		"embedded pointer",
+		`type S14 struct { *S10; b int }
+		func F%d() int {
+			var s S14
+			s.S10 = &S10{}
+			s.a = 31
+			s.b = 11
+			return s.a + s.b
+		}
+		`,
+		big.NewInt(42),
+	},
+	{
+		"same field names",
+		`type S15 struct { S10; a int; b int }
+		func F%d() int {
+			var s S15
+			s.S10.a = 31
+			s.a = 5
+			s.b = 6
+			return s.S10.a + s.a + s.b
+		}
+		`,
+		big.NewInt(42),
+	},
+	{
+		"embedded pointer chain",
+		`type S16 struct { *S10 }
+		type S17 struct { *S16; b int }
+		func F%d() int {
+			var c S17
+			c.S16 = &S16{S10: &S10{a: 31}}
+			c.b = 11
+			return c.a + c.b
+		}
+		`,
+		big.NewInt(42),
+	},
 }
 
 func TestStructs(t *testing.T) {

